File: Pintpal_web/src/features/ratings/data/pintPhotoRepository.ts
/**
 * pintPhotoRepository.ts
 *
 * Purpose: Upload, list, rate Best Pints photos; daily/weekly winners + votes.
 * Connects to: Firestore pint_photos, contest_weeks, contest_days + Storage uploads.
 */

import {
  collection,
  doc,
  getDoc,
  getDocs,
  limit,
  orderBy,
  query,
  runTransaction,
  setDoc,
  where,
  type DocumentData,
} from "firebase/firestore";
import {
  getDownloadURL,
  ref,
  uploadBytes,
} from "firebase/storage";
import { getFirebaseAuth, getFirebaseFirestore, getFirebaseStorage } from "@/utilities/firebase";
import {
  Collections,
  ContestDayFields,
  ContestVoteFields,
  ContestWeekFields,
  PintPhotoFields,
  PintPhotoRatingFields,
  UserFields,
} from "@/utilities/firebaseConstants";
import { loadUserDocument } from "@/features/auth/data/authRepository";
import {
  canUploadPhoto,
  isSubscriptionPaid,
  photoUploadsUsedToday,
} from "@/utilities/permissions";
import {
  dayKeyFromDate,
  previousWeekKey,
  weekKeyFromDate,
  yesterdayDayKey,
} from "./contestKeys";
import type { ContestDay, ContestWeek, PintPhoto } from "../types";

function todayUtc(): string {
  return dayKeyFromDate();
}

function mapPhoto(id: string, data: DocumentData): PintPhoto {
  return {
    id,
    userId: typeof data[PintPhotoFields.userId] === "string" ? data[PintPhotoFields.userId] : "",
    displayName:
      typeof data[PintPhotoFields.displayName] === "string"
        ? data[PintPhotoFields.displayName]
        : "Member",
    imageUrl:
      typeof data[PintPhotoFields.imageUrl] === "string"
        ? data[PintPhotoFields.imageUrl]
        : "",
    storagePath:
      typeof data[PintPhotoFields.storagePath] === "string"
        ? data[PintPhotoFields.storagePath]
        : "",
    createdAtMs:
      typeof data[PintPhotoFields.createdAt] === "number"
        ? data[PintPhotoFields.createdAt]
        : 0,
    dayKey:
      typeof data[PintPhotoFields.dayKey] === "string"
        ? data[PintPhotoFields.dayKey]
        : "",
    weekKey:
      typeof data[PintPhotoFields.weekKey] === "string"
        ? data[PintPhotoFields.weekKey]
        : "",
    ratingSum:
      typeof data[PintPhotoFields.ratingSum] === "number"
        ? data[PintPhotoFields.ratingSum]
        : 0,
    ratingCount:
      typeof data[PintPhotoFields.ratingCount] === "number"
        ? data[PintPhotoFields.ratingCount]
        : 0,
    averageRating:
      typeof data[PintPhotoFields.averageRating] === "number"
        ? data[PintPhotoFields.averageRating]
        : 0,
  };
}

function sortByAverage(photos: PintPhoto[]): PintPhoto[] {
  return [...photos].sort((a, b) => {
    if (b.averageRating !== a.averageRating) {
      return b.averageRating - a.averageRating;
    }
    return b.ratingCount - a.ratingCount;
  });
}

export async function listRecentPhotos(max = 40): Promise<PintPhoto[]> {
  const db = getFirebaseFirestore();
  const q = query(
    collection(db, Collections.pintPhotos),
    orderBy(PintPhotoFields.createdAt, "desc"),
    limit(max),
  );
  const snap = await getDocs(q);
  return snap.docs.map((d) => mapPhoto(d.id, d.data()));
}

export async function listPhotosForDay(dayKey: string): Promise<PintPhoto[]> {
  const db = getFirebaseFirestore();
  const q = query(
    collection(db, Collections.pintPhotos),
    where(PintPhotoFields.dayKey, "==", dayKey),
    limit(100),
  );
  const snap = await getDocs(q);
  return sortByAverage(snap.docs.map((d) => mapPhoto(d.id, d.data())));
}

export async function listPhotosForWeek(weekKey: string): Promise<PintPhoto[]> {
  const db = getFirebaseFirestore();
  const q = query(
    collection(db, Collections.pintPhotos),
    where(PintPhotoFields.weekKey, "==", weekKey),
    limit(100),
  );
  const snap = await getDocs(q);
  return sortByAverage(snap.docs.map((d) => mapPhoto(d.id, d.data())));
}

export async function uploadPintPhoto(file: File): Promise<PintPhoto> {
  const auth = getFirebaseAuth();
  const user = auth.currentUser;
  if (!user) throw new Error("Not signed in");

  const profile = await loadUserDocument(user.uid);
  if (!profile) throw new Error("User profile missing");
  if (!canUploadPhoto(profile)) {
    throw new Error("Photo upload limit reached for today (free plan).");
  }

  const dayKey = todayUtc();
  const weekKey = weekKeyFromDate();
  const storagePath = `users/${user.uid}/uploads/pint_${Date.now()}_${file.name.replace(/[^a-zA-Z0-9._-]/g, "_")}`;
  const storageRef = ref(getFirebaseStorage(), storagePath);
  await uploadBytes(storageRef, file, { contentType: file.type || "image/jpeg" });
  const imageUrl = await getDownloadURL(storageRef);

  const db = getFirebaseFirestore();
  const photoRef = doc(collection(db, Collections.pintPhotos));
  const createdAt = Date.now();
  const payload = {
    [PintPhotoFields.userId]: user.uid,
    [PintPhotoFields.displayName]: profile.name || profile.email || "Member",
    [PintPhotoFields.imageUrl]: imageUrl,
    [PintPhotoFields.storagePath]: storagePath,
    [PintPhotoFields.createdAt]: createdAt,
    [PintPhotoFields.dayKey]: dayKey,
    [PintPhotoFields.weekKey]: weekKey,
    [PintPhotoFields.ratingSum]: 0,
    [PintPhotoFields.ratingCount]: 0,
    [PintPhotoFields.averageRating]: 0,
  };
  await setDoc(photoRef, payload);

  // Free-tier quota counter on users/{uid}.
  if (!isSubscriptionPaid(profile)) {
    const used = photoUploadsUsedToday(profile);
    await setDoc(
      doc(db, Collections.users, user.uid),
      {
        [UserFields.photoUploadsToday]: used + 1,
        [UserFields.photoUploadDate]: dayKey,
      },
      { merge: true },
    );
  }

  return mapPhoto(photoRef.id, payload);
}

export async function ratePhoto(photoId: string, score: number): Promise<void> {
  const auth = getFirebaseAuth();
  const user = auth.currentUser;
  if (!user) throw new Error("Not signed in");
  const clamped = Math.round(Math.min(10, Math.max(1, score)));

  const db = getFirebaseFirestore();
  const photoRef = doc(db, Collections.pintPhotos, photoId);
  const ratingRef = doc(photoRef, "ratings", user.uid);

  await runTransaction(db, async (tx) => {
    const photoSnap = await tx.get(photoRef);
    if (!photoSnap.exists()) throw new Error("Photo not found");
    const data = photoSnap.data();
    const ratingSnap = await tx.get(ratingRef);
    const oldSum =
      typeof data[PintPhotoFields.ratingSum] === "number"
        ? data[PintPhotoFields.ratingSum]
        : 0;
    const oldCount =
      typeof data[PintPhotoFields.ratingCount] === "number"
        ? data[PintPhotoFields.ratingCount]
        : 0;

    let newSum = oldSum;
    let newCount = oldCount;
    if (ratingSnap.exists()) {
      const prev =
        typeof ratingSnap.data()[PintPhotoRatingFields.score] === "number"
          ? ratingSnap.data()[PintPhotoRatingFields.score]
          : 0;
      newSum = oldSum - prev + clamped;
    } else {
      newSum = oldSum + clamped;
      newCount = oldCount + 1;
    }
    const average = newCount > 0 ? newSum / newCount : 0;
    tx.set(ratingRef, {
      [PintPhotoRatingFields.score]: clamped,
      [PintPhotoRatingFields.createdAt]: Date.now(),
    });
    tx.update(photoRef, {
      [PintPhotoFields.ratingSum]: newSum,
      [PintPhotoFields.ratingCount]: newCount,
      [PintPhotoFields.averageRating]: average,
    });
  });
}

export async function ensureDailyWinner(dayKey: string): Promise<ContestDay | null> {
  const ranked = await listPhotosForDay(dayKey);
  const top = ranked.find((p) => p.ratingCount > 0) ?? ranked[0];
  if (!top) return null;

  const db = getFirebaseFirestore();
  const dayRef = doc(db, Collections.contestDays, dayKey);
  const payload = {
    [ContestDayFields.winnerId]: top.id,
    [ContestDayFields.averageRating]: top.averageRating,
    [ContestDayFields.displayName]: top.displayName,
    [ContestDayFields.imageUrl]: top.imageUrl,
    [ContestDayFields.updatedAt]: Date.now(),
  };
  await setDoc(dayRef, payload, { merge: true });
  return {
    dayKey,
    winnerId: top.id,
    averageRating: top.averageRating,
    displayName: top.displayName,
    imageUrl: top.imageUrl,
  };
}

export async function loadDailyWinner(dayKey: string): Promise<ContestDay | null> {
  const db = getFirebaseFirestore();
  const snap = await getDoc(doc(db, Collections.contestDays, dayKey));
  if (!snap.exists()) return ensureDailyWinner(dayKey);
  const data = snap.data();
  return {
    dayKey,
    winnerId: String(data[ContestDayFields.winnerId] ?? ""),
    averageRating:
      typeof data[ContestDayFields.averageRating] === "number"
        ? data[ContestDayFields.averageRating]
        : 0,
    displayName: String(data[ContestDayFields.displayName] ?? ""),
    imageUrl: String(data[ContestDayFields.imageUrl] ?? ""),
  };
}

export async function ensureWeeklyFinalists(
  weekKey: string,
): Promise<ContestWeek> {
  const db = getFirebaseFirestore();
  const weekRef = doc(db, Collections.contestWeeks, weekKey);
  const existing = await getDoc(weekRef);
  if (existing.exists()) {
    const data = existing.data();
    const ids = Array.isArray(data[ContestWeekFields.finalistIds])
      ? (data[ContestWeekFields.finalistIds] as string[]).slice(0, 3)
      : [];
    return {
      weekKey,
      status: (data[ContestWeekFields.status] as ContestWeek["status"]) || "voting",
      finalistIds: ids,
      winnerId:
        typeof data[ContestWeekFields.winnerId] === "string"
          ? data[ContestWeekFields.winnerId]
          : null,
    };
  }

  const ranked = await listPhotosForWeek(weekKey);
  const finalistIds = ranked
    .filter((p) => p.ratingCount > 0)
    .slice(0, 3)
    .map((p) => p.id);
  // If no ratings yet, still take top by presence.
  const ids =
    finalistIds.length > 0
      ? finalistIds
      : ranked.slice(0, 3).map((p) => p.id);

  const payload = {
    [ContestWeekFields.status]: "voting",
    [ContestWeekFields.finalistIds]: ids,
    [ContestWeekFields.winnerId]: null,
    [ContestWeekFields.updatedAt]: Date.now(),
  };
  await setDoc(weekRef, payload);
  return {
    weekKey,
    status: "voting",
    finalistIds: ids,
    winnerId: null,
  };
}

export async function loadPhotosByIds(ids: string[]): Promise<PintPhoto[]> {
  if (ids.length === 0) return [];
  const db = getFirebaseFirestore();
  const out: PintPhoto[] = [];
  for (const id of ids) {
    const snap = await getDoc(doc(db, Collections.pintPhotos, id));
    if (snap.exists()) out.push(mapPhoto(snap.id, snap.data()));
  }
  return out;
}

export async function voteWeeklyWinner(
  weekKey: string,
  photoId: string,
): Promise<void> {
  const auth = getFirebaseAuth();
  const user = auth.currentUser;
  if (!user) throw new Error("Not signed in");

  const week = await ensureWeeklyFinalists(weekKey);
  if (!week.finalistIds.includes(photoId)) {
    throw new Error("That photo is not in this week's top 3.");
  }

  const db = getFirebaseFirestore();
  const voteRef = doc(db, Collections.contestWeeks, weekKey, "votes", user.uid);
  await setDoc(voteRef, {
    [ContestVoteFields.photoId]: photoId,
    [ContestVoteFields.createdAt]: Date.now(),
  });

  // Tally votes and set winner.
  const votesSnap = await getDocs(
    collection(db, Collections.contestWeeks, weekKey, "votes"),
  );
  const tallies = new Map<string, number>();
  for (const v of votesSnap.docs) {
    const pid = v.data()[ContestVoteFields.photoId];
    if (typeof pid === "string") {
      tallies.set(pid, (tallies.get(pid) ?? 0) + 1);
    }
  }
  let winnerId: string | null = null;
  let best = -1;
  for (const id of week.finalistIds) {
    const n = tallies.get(id) ?? 0;
    if (n > best) {
      best = n;
      winnerId = id;
    }
  }
  if (winnerId) {
    await setDoc(
      doc(db, Collections.contestWeeks, weekKey),
      {
        [ContestWeekFields.winnerId]: winnerId,
        [ContestWeekFields.status]: "voting",
        [ContestWeekFields.updatedAt]: Date.now(),
      },
      { merge: true },
    );
  }
}

export async function loadContestSnapshot(): Promise<{
  recent: PintPhoto[];
  todayLeader: ContestDay | null;
  yesterdayWinner: ContestDay | null;
  thisWeekTop3: PintPhoto[];
  lastWeek: ContestWeek;
  lastWeekFinalists: PintPhoto[];
  currentWeekKey: string;
  previousWeekKey: string;
}> {
  const currentWeek = weekKeyFromDate();
  const prevWeek = previousWeekKey();
  const today = dayKeyFromDate();
  const yesterday = yesterdayDayKey();

  const [recent, thisWeek, todayLeader, yesterdayWinner, lastWeek] =
    await Promise.all([
      listRecentPhotos(),
      listPhotosForWeek(currentWeek),
      ensureDailyWinner(today),
      loadDailyWinner(yesterday),
      ensureWeeklyFinalists(prevWeek),
    ]);

  const lastWeekFinalists = await loadPhotosByIds(lastWeek.finalistIds);

  return {
    recent,
    todayLeader,
    yesterdayWinner,
    thisWeekTop3: thisWeek.filter((p) => p.ratingCount > 0).slice(0, 3),
    lastWeek,
    lastWeekFinalists,
    currentWeekKey: currentWeek,
    previousWeekKey: prevWeek,
  };
}

File: pintpal_app/app/src/main/java/com/jdscott/pintpal/features/ratings/data/PintPhotoRepository.kt
/**
 * PintPhotoRepository.kt
 *
 * Purpose: Upload, list, rate Best Pints photos; daily/weekly winners + votes.
 */
package com.jdscott.pintpal.features.ratings.data

import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.Query
import com.google.firebase.firestore.SetOptions
import com.google.firebase.storage.FirebaseStorage
import com.jdscott.pintpal.features.auth.data.AuthRepository
import com.jdscott.pintpal.utilities.FirebaseConstants
import com.jdscott.pintpal.utilities.UserPermissions
import kotlinx.coroutines.tasks.await

data class PintPhoto(
    val id: String,
    val userId: String,
    val displayName: String,
    val imageUrl: String,
    val storagePath: String,
    val createdAtMs: Long,
    val dayKey: String,
    val weekKey: String,
    val ratingSum: Double,
    val ratingCount: Int,
    val averageRating: Double,
)

data class ContestDay(
    val dayKey: String,
    val winnerId: String,
    val averageRating: Double,
    val displayName: String,
    val imageUrl: String,
)

data class ContestWeek(
    val weekKey: String,
    val status: String,
    val finalistIds: List<String>,
    val winnerId: String?,
)

data class ContestSnapshot(
    val recent: List<PintPhoto>,
    val todayLeader: ContestDay?,
    val yesterdayWinner: ContestDay?,
    val thisWeekTop3: List<PintPhoto>,
    val lastWeek: ContestWeek,
    val lastWeekFinalists: List<PintPhoto>,
    val currentWeekKey: String,
    val previousWeekKey: String,
)

class PintPhotoRepository(
    private val auth: FirebaseAuth = FirebaseAuth.getInstance(),
    private val db: FirebaseFirestore = FirebaseFirestore.getInstance(),
    private val storage: FirebaseStorage = FirebaseStorage.getInstance(),
    private val authRepository: AuthRepository = AuthRepository(),
) {
    private fun mapPhoto(id: String, data: Map<String, Any?>): PintPhoto {
        fun num(key: String): Double =
            (data[key] as? Number)?.toDouble() ?: 0.0
        fun int(key: String): Int =
            (data[key] as? Number)?.toInt() ?: 0
        fun str(key: String): String =
            data[key] as? String ?: ""

        return PintPhoto(
            id = id,
            userId = str(FirebaseConstants.PintPhotoFields.USER_ID),
            displayName = str(FirebaseConstants.PintPhotoFields.DISPLAY_NAME).ifBlank { "Member" },
            imageUrl = str(FirebaseConstants.PintPhotoFields.IMAGE_URL),
            storagePath = str(FirebaseConstants.PintPhotoFields.STORAGE_PATH),
            createdAtMs = (data[FirebaseConstants.PintPhotoFields.CREATED_AT] as? Number)?.toLong() ?: 0L,
            dayKey = str(FirebaseConstants.PintPhotoFields.DAY_KEY),
            weekKey = str(FirebaseConstants.PintPhotoFields.WEEK_KEY),
            ratingSum = num(FirebaseConstants.PintPhotoFields.RATING_SUM),
            ratingCount = int(FirebaseConstants.PintPhotoFields.RATING_COUNT),
            averageRating = num(FirebaseConstants.PintPhotoFields.AVERAGE_RATING),
        )
    }

    private fun sortByAverage(photos: List<PintPhoto>): List<PintPhoto> =
        photos.sortedWith(
            compareByDescending<PintPhoto> { it.averageRating }
                .thenByDescending { it.ratingCount },
        )

    suspend fun listRecentPhotos(max: Long = 40): List<PintPhoto> {
        val snap = db.collection(FirebaseConstants.Collections.PINT_PHOTOS)
            .orderBy(FirebaseConstants.PintPhotoFields.CREATED_AT, Query.Direction.DESCENDING)
            .limit(max)
            .get()
            .await()
        return snap.documents.map { mapPhoto(it.id, it.data ?: emptyMap()) }
    }

    suspend fun listPhotosForDay(dayKey: String): List<PintPhoto> {
        val snap = db.collection(FirebaseConstants.Collections.PINT_PHOTOS)
            .whereEqualTo(FirebaseConstants.PintPhotoFields.DAY_KEY, dayKey)
            .limit(100)
            .get()
            .await()
        return sortByAverage(snap.documents.map { mapPhoto(it.id, it.data ?: emptyMap()) })
    }

    suspend fun listPhotosForWeek(weekKey: String): List<PintPhoto> {
        val snap = db.collection(FirebaseConstants.Collections.PINT_PHOTOS)
            .whereEqualTo(FirebaseConstants.PintPhotoFields.WEEK_KEY, weekKey)
            .limit(100)
            .get()
            .await()
        return sortByAverage(snap.documents.map { mapPhoto(it.id, it.data ?: emptyMap()) })
    }

    suspend fun uploadPintPhoto(bytes: ByteArray, contentType: String, fileName: String): PintPhoto {
        val user = auth.currentUser ?: error("Not signed in")
        val profile = authRepository.loadCurrentUserDocument().getOrNull()
            ?: error("User profile missing")
        val flags = profile.toPermissionFlags()
        if (!UserPermissions.canUploadPhoto(flags)) {
            error("Photo upload limit reached for today (free plan).")
        }

        val dayKey = ContestKeys.dayKey()
        val weekKey = ContestKeys.weekKey()
        val safeName = fileName.replace(Regex("[^a-zA-Z0-9._-]"), "_")
        val storagePath = "users/${user.uid}/uploads/pint_${System.currentTimeMillis()}_$safeName"
        val storageRef = storage.reference.child(storagePath)
        storageRef.putBytes(bytes, com.google.firebase.storage.StorageMetadata.Builder()
            .setContentType(contentType.ifBlank { "image/jpeg" })
            .build()).await()
        val imageUrl = storageRef.downloadUrl.await().toString()

        val photoRef = db.collection(FirebaseConstants.Collections.PINT_PHOTOS).document()
        val createdAt = System.currentTimeMillis()
        val payload = hashMapOf<String, Any>(
            FirebaseConstants.PintPhotoFields.USER_ID to user.uid,
            FirebaseConstants.PintPhotoFields.DISPLAY_NAME to
                profile.name.ifBlank { profile.email.ifBlank { "Member" } },
            FirebaseConstants.PintPhotoFields.IMAGE_URL to imageUrl,
            FirebaseConstants.PintPhotoFields.STORAGE_PATH to storagePath,
            FirebaseConstants.PintPhotoFields.CREATED_AT to createdAt,
            FirebaseConstants.PintPhotoFields.DAY_KEY to dayKey,
            FirebaseConstants.PintPhotoFields.WEEK_KEY to weekKey,
            FirebaseConstants.PintPhotoFields.RATING_SUM to 0,
            FirebaseConstants.PintPhotoFields.RATING_COUNT to 0,
            FirebaseConstants.PintPhotoFields.AVERAGE_RATING to 0,
        )
        photoRef.set(payload).await()

        if (!UserPermissions.isSubscriptionPaid(flags)) {
            val used = UserPermissions.photoUploadsUsedToday(flags)
            db.collection(FirebaseConstants.Collections.USERS).document(user.uid)
                .set(
                    mapOf(
                        FirebaseConstants.UserFields.PHOTO_UPLOADS_TODAY to used + 1,
                        FirebaseConstants.UserFields.PHOTO_UPLOAD_DATE to dayKey,
                    ),
                    SetOptions.merge(),
                )
                .await()
        }

        return mapPhoto(photoRef.id, payload)
    }

    suspend fun ratePhoto(photoId: String, score: Int) {
        val user = auth.currentUser ?: error("Not signed in")
        val clamped = score.coerceIn(1, 10)
        val photoRef = db.collection(FirebaseConstants.Collections.PINT_PHOTOS).document(photoId)
        val ratingRef = photoRef.collection("ratings").document(user.uid)

        db.runTransaction { tx ->
            val photoSnap = tx.get(photoRef)
            if (!photoSnap.exists()) error("Photo not found")
            val ratingSnap = tx.get(ratingRef)
            val oldSum = (photoSnap.get(FirebaseConstants.PintPhotoFields.RATING_SUM) as? Number)?.toDouble() ?: 0.0
            val oldCount = (photoSnap.get(FirebaseConstants.PintPhotoFields.RATING_COUNT) as? Number)?.toInt() ?: 0
            var newSum = oldSum
            var newCount = oldCount
            if (ratingSnap.exists()) {
                val prev = (ratingSnap.get(FirebaseConstants.PintPhotoRatingFields.SCORE) as? Number)?.toDouble() ?: 0.0
                newSum = oldSum - prev + clamped
            } else {
                newSum = oldSum + clamped
                newCount = oldCount + 1
            }
            val average = if (newCount > 0) newSum / newCount else 0.0
            tx.set(
                ratingRef,
                mapOf(
                    FirebaseConstants.PintPhotoRatingFields.SCORE to clamped,
                    FirebaseConstants.PintPhotoRatingFields.CREATED_AT to System.currentTimeMillis(),
                ),
            )
            tx.update(
                photoRef,
                mapOf(
                    FirebaseConstants.PintPhotoFields.RATING_SUM to newSum,
                    FirebaseConstants.PintPhotoFields.RATING_COUNT to newCount,
                    FirebaseConstants.PintPhotoFields.AVERAGE_RATING to average,
                ),
            )
        }.await()
    }

    suspend fun ensureDailyWinner(dayKey: String): ContestDay? {
        val ranked = listPhotosForDay(dayKey)
        val top = ranked.firstOrNull { it.ratingCount > 0 } ?: ranked.firstOrNull() ?: return null
        val payload = mapOf(
            FirebaseConstants.ContestDayFields.WINNER_ID to top.id,
            FirebaseConstants.ContestDayFields.AVERAGE_RATING to top.averageRating,
            FirebaseConstants.ContestDayFields.DISPLAY_NAME to top.displayName,
            FirebaseConstants.ContestDayFields.IMAGE_URL to top.imageUrl,
            FirebaseConstants.ContestDayFields.UPDATED_AT to System.currentTimeMillis(),
        )
        db.collection(FirebaseConstants.Collections.CONTEST_DAYS).document(dayKey)
            .set(payload, SetOptions.merge())
            .await()
        return ContestDay(
            dayKey = dayKey,
            winnerId = top.id,
            averageRating = top.averageRating,
            displayName = top.displayName,
            imageUrl = top.imageUrl,
        )
    }

    suspend fun loadDailyWinner(dayKey: String): ContestDay? {
        val snap = db.collection(FirebaseConstants.Collections.CONTEST_DAYS).document(dayKey).get().await()
        if (!snap.exists()) return ensureDailyWinner(dayKey)
        return ContestDay(
            dayKey = dayKey,
            winnerId = snap.getString(FirebaseConstants.ContestDayFields.WINNER_ID).orEmpty(),
            averageRating = (snap.get(FirebaseConstants.ContestDayFields.AVERAGE_RATING) as? Number)?.toDouble() ?: 0.0,
            displayName = snap.getString(FirebaseConstants.ContestDayFields.DISPLAY_NAME).orEmpty(),
            imageUrl = snap.getString(FirebaseConstants.ContestDayFields.IMAGE_URL).orEmpty(),
        )
    }

    suspend fun ensureWeeklyFinalists(weekKey: String): ContestWeek {
        val weekRef = db.collection(FirebaseConstants.Collections.CONTEST_WEEKS).document(weekKey)
        val existing = weekRef.get().await()
        if (existing.exists()) {
            @Suppress("UNCHECKED_CAST")
            val ids = (existing.get(FirebaseConstants.ContestWeekFields.FINALIST_IDS) as? List<*>)
                ?.mapNotNull { it as? String }
                ?.take(3)
                .orEmpty()
            return ContestWeek(
                weekKey = weekKey,
                status = existing.getString(FirebaseConstants.ContestWeekFields.STATUS) ?: "voting",
                finalistIds = ids,
                winnerId = existing.getString(FirebaseConstants.ContestWeekFields.WINNER_ID),
            )
        }
        val ranked = listPhotosForWeek(weekKey)
        val rated = ranked.filter { it.ratingCount > 0 }.take(3).map { it.id }
        val ids = if (rated.isNotEmpty()) rated else ranked.take(3).map { it.id }
        weekRef.set(
            mapOf(
                FirebaseConstants.ContestWeekFields.STATUS to "voting",
                FirebaseConstants.ContestWeekFields.FINALIST_IDS to ids,
                FirebaseConstants.ContestWeekFields.WINNER_ID to null,
                FirebaseConstants.ContestWeekFields.UPDATED_AT to System.currentTimeMillis(),
            ),
        ).await()
        return ContestWeek(weekKey, "voting", ids, null)
    }

    suspend fun loadPhotosByIds(ids: List<String>): List<PintPhoto> {
        if (ids.isEmpty()) return emptyList()
        return ids.mapNotNull { id ->
            val snap = db.collection(FirebaseConstants.Collections.PINT_PHOTOS).document(id).get().await()
            if (!snap.exists()) null else mapPhoto(snap.id, snap.data ?: emptyMap())
        }
    }

    suspend fun voteWeeklyWinner(weekKey: String, photoId: String) {
        val user = auth.currentUser ?: error("Not signed in")
        val week = ensureWeeklyFinalists(weekKey)
        if (photoId !in week.finalistIds) error("That photo is not in this week's top 3.")

        val voteRef = db.collection(FirebaseConstants.Collections.CONTEST_WEEKS)
            .document(weekKey)
            .collection("votes")
            .document(user.uid)
        voteRef.set(
            mapOf(
                FirebaseConstants.ContestVoteFields.PHOTO_ID to photoId,
                FirebaseConstants.ContestVoteFields.CREATED_AT to System.currentTimeMillis(),
            ),
        ).await()

        val votes = db.collection(FirebaseConstants.Collections.CONTEST_WEEKS)
            .document(weekKey)
            .collection("votes")
            .get()
            .await()
        val tallies = mutableMapOf<String, Int>()
        for (doc in votes.documents) {
            val pid = doc.getString(FirebaseConstants.ContestVoteFields.PHOTO_ID) ?: continue
            tallies[pid] = (tallies[pid] ?: 0) + 1
        }
        var winnerId: String? = null
        var best = -1
        for (id in week.finalistIds) {
            val n = tallies[id] ?: 0
            if (n > best) {
                best = n
                winnerId = id
            }
        }
        if (winnerId != null) {
            db.collection(FirebaseConstants.Collections.CONTEST_WEEKS).document(weekKey)
                .set(
                    mapOf(
                        FirebaseConstants.ContestWeekFields.WINNER_ID to winnerId,
                        FirebaseConstants.ContestWeekFields.STATUS to "voting",
                        FirebaseConstants.ContestWeekFields.UPDATED_AT to System.currentTimeMillis(),
                    ),
                    SetOptions.merge(),
                )
                .await()
        }
    }

    suspend fun loadContestSnapshot(): ContestSnapshot {
        val currentWeek = ContestKeys.weekKey()
        val prevWeek = ContestKeys.previousWeekKey()
        val today = ContestKeys.dayKey()
        val yesterday = ContestKeys.yesterdayDayKey()

        val recent = listRecentPhotos()
        val thisWeek = listPhotosForWeek(currentWeek)
        val todayLeader = ensureDailyWinner(today)
        val yesterdayWinner = loadDailyWinner(yesterday)
        val lastWeek = ensureWeeklyFinalists(prevWeek)
        val lastWeekFinalists = loadPhotosByIds(lastWeek.finalistIds)

        return ContestSnapshot(
            recent = recent,
            todayLeader = todayLeader,
            yesterdayWinner = yesterdayWinner,
            thisWeekTop3 = thisWeek.filter { it.ratingCount > 0 }.take(3),
            lastWeek = lastWeek,
            lastWeekFinalists = lastWeekFinalists,
            currentWeekKey = currentWeek,
            previousWeekKey = prevWeek,
        )
    }
}
